### PR TITLE
adding new phase in competition fixed

### LINKED ISSUE #741: Adding another phase to pre-existing competition fails
Creating a competition from scratch allows multiple phases, but adding one during an "update" is not working and produces an error.

image

Phases a and b were created upon challenge creation and attempting to add c is failing with error.

### PR DESCRIPTION
**Pull request by ihsaan-ullah**
# @ mention of reviewers
@Didayolo 

# A brief description of the purpose of the changes contained in this PR.
Now users can add new phase in a competition

# Issues this PR resolves
- #741 

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [x] Code review by reviewer
- [x] Hand tested by reviewer
- [x] CircleCi tests are passing
- [x] Ready to merge

### CODE REVIEW

File: src/apps/api/views/competitions.py [hunk 1]
```diff
@@ -201,7 +202,32 @@ def update(self, request, *args, **kwargs):
...
             leaderboard_id = leaderboard["id"].value
+
             for phase in data['phases']:
+                # Newly added phase from front-end has no id
+                # Add a phase first to get id
+                # add this phase id in each task
+                if 'id' not in phase:
+
+                    # Create Phase object
+                    # TODO: wrap this whole phase addition inside competition updation in a Transaction in case something fails after the phase id is created and not yet linked to tasks
```
Comment (Didayolo) [inline src/apps/api/views/competitions.py:213]: It is normal that you added a TODO?

**Comment by Didayolo** [comment]
But now we have this error in CircleCI:

```python
E               django.db.utils.IntegrityError: duplicate key value violates unique constraint "leaderboards_column_leaderboard_id_key_aa81662d_uniq"
E               DETAIL:  Key (leaderboard_id, key)=(7, nature) already exists.

/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py:84: IntegrityError
```

```python
=========================== short test summary info ============================
FAILED src/apps/api/tests/test_competitions.py::CompetitionResultDatatypesTests::test_get_competition_leaderboard_as_json
===== 1 failed, 167 passed, 9 deselected, 89 warnings in 84.84s (0:01:24) ======
```

**Comment by ihsaan-ullah** [comment]
I don't know why it failed but rerunning the test fixed the failure :D

**Comment by ihsaan-ullah** [comment]
To check

- [ ] when a competition is deleted, its phases are also deleted?

File: src/apps/api/views/competitions.py (hunk 1 above)
Comment (ihsaan-ullah) [reply src/apps/api/views/competitions.py]: added transaction

**Comment by Didayolo** [comment]
Depending on the order and dates of phase, I still have the same problem: nothing happens when I click on save.

### First test

I put the newly created phase as the oldest (top of the list), with start and end dates before the next phase. It saves and work fine.

### Second test

I put the newly created phase as the newest (bottom of the list), with a starting date and no end date. I edit the dates of the other phases so they start and finish before the newly created phase. This DOES NOT save.

**Comment by Didayolo** [comment]
> I don't know why it failed but rerunning the test fixed the failure :D

Isn't it a bit concerning that it failed in the first place?

**Comment by ihsaan-ullah** [comment]
More tests:

I have a competition with 2 phases:  `Phase 1` and `Phase 2`
Screenshot 2023-06-22 at 8 25 47 PM

Screenshot 2023-06-22 at 8 25 59 PM

`Phase 1` : 01 Jan 2023 - 31 Jan 2023. 
`Phase 2` : 01 Feb 2023 - 28 Feb 2023. 

Adding two phases to this competiiton: one before phase 1 and one after phase 2

`Phase 0` 01 Dec 2022 - 31 Dec 2022. ✅   
Screenshot 2023-06-22 at 8 29 55 PM

`Phase 3` 01 Mar 2023 - 31 Mar 2023. ✅  
Screenshot 2023-06-22 at 8 31 29 PM


@Didayolo I think everything is working. Can you try once more with a fresh branch and with collectstatic.

**Comment by Didayolo** [comment]
In my case, the last phase was "never ending" before I change the dates. Maybe the problem comes from that? It was a fresh deployement with collect static.

**Comment by ihsaan-ullah** [comment]
> In my case, the last phase was "never ending" before I changed the dates. Maybe the problem comes from that? It was a fresh deployement with collect static.

i will check this condition too

**Comment by ihsaan-ullah** [comment]
@Didayolo Final phase has no end date. And when i add a phase after that, I get this error which seems reasonable. What do you think?

Screenshot 2023-06-22 at 9 55 55 PM

**Comment by Didayolo** [comment]
>  Final phase has no end date. And when i add a phase after that, I get this error which seems reasonable. What do you think?

OK, so I was exactly in this situation, and updated `Final phase` and `Final phase (new addition)` dates to remove the warning:
- Put an end date to `Final phase`
- Put a start date to `Final phase (new addition)` and no end date

So the warning disappear but then the `Save` button does not work.

**Comment by ihsaan-ullah** [comment]
@Didayolo that was a nice catch. Now last phase can have no end date.

**Comment by Didayolo** [comment]
The fix is a nice catch too!

**Pull request merged**